source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+#  if 0
+    PetscPrintf(PETSC_COMM_WORLD, "INIDX\n");
+    PetscIntView(inidx.size(), inidx.data(), PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "INLOC\n");
+    PetscIntView(inloc.size(),
+                 inloc.size() > 0 ? inloc.data() : nullptr,
+                 PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "OUTIDX\n");
+    PetscIntView(outidx.size(), outidx.data(), PETSC_VIEWER_STDOUT_WORLD);
+    PetscPrintf(PETSC_COMM_WORLD, "OUTLOC\n");
+    PetscIntView(outloc.size(),
+                 outloc.size() > 0 ? outloc.data() : nullptr,
+                 PETSC_VIEWER_STDOUT_WORLD);
+    if (PetscGlobalRank == 0)
+      {
+        PetscPrintf(PETSC_COMM_WORLD, "RANGES\n");
+        PetscMPIInt size;
+        MPI_Comm_size(communicator, &size);
+        PetscIntView(size + 1, ranges, PETSC_VIEWER_STDOUT_SELF);
+      }
+#  endif
+    PetscInt    cnt   = 0;
+    PetscMPIInt owner = 0;
+    for (const auto idx : inidx)
+      {
+        // short-circuit the search if the last owner owns this index too
+        if (idx < ranges[owner] || ranges[owner + 1] <= idx)
+          {
+            AssertPETSc(PetscLayoutFindOwner(layout, idx, &owner));
+          }
+        remotes[cnt].rank  = owner;
+        remotes[cnt].index = idx - ranges[owner];
+        cnt++;
+      }
+
+    AssertPETSc(PetscSFCreate(communicator, &sf2));
+    AssertPETSc(PetscSFSetGraph(sf2,
+                                nl,
+                                n,
+                                const_cast<PetscInt *>(
+                                  inloc.size() > 0 ? inloc.data() : nullptr),
+                                PETSC_COPY_VALUES,
+                                remotes,
+                                PETSC_OWN_POINTER));
+    AssertPETSc(PetscSFSetUp(sf2));
+    // AssertPETSc(PetscSFView(sf2, nullptr));
+    // We need to invert root and leaf space to create the first SF
+    AssertPETSc(PetscSFCreateInverseSF(sf2, &sf1));
+    AssertPETSc(PetscSFDestroy(&sf2));
+    // AssertPETSc(PetscSFView(sf1, nullptr));
+
+    // Now create the SF from the contiguous space to the local output space
+    n = static_cast<PetscInt>(outidx.size());
+    AssertPETSc(PetscSFCreate(communicator, &sf2));
+    AssertPETSc(PetscSFSetGraphLayout(
+      sf2,
+      layout,
+      n,
+      const_cast<PetscInt *>(outloc.size() > 0 ? outloc.data() : nullptr),
+      PETSC_COPY_VALUES,
+      const_cast<PetscInt *>(n > 0 ? outidx.data() : nullptr)));
+    AssertPETSc(PetscSFSetUp(sf2));
+    // AssertPETSc(PetscSFView(sf2, nullptr));
+
+    // The final SF is the composition of the two
+    AssertPETSc(PetscSFCompose(sf1, sf2, &sf));
+    // AssertPETSc(PetscSFView(sf, nullptr));
+
+    // Cleanup
+    AssertPETSc(PetscLayoutDestroy(&layout));
+    AssertPETSc(PetscSFDestroy(&sf1));
+    AssertPETSc(PetscSFDestroy(&sf2));
+  }
+
+  void
+  CommunicationPattern::clear()
+  {
+    AssertPETSc(PetscSFDestroy(&sf));
+  }
+
+  const MPI_Comm &
```

Comment:
Here you have an idea of how we need to implement `get_mpi_communicator()` if we keep the reference in the return argument. We cannot use the static variable inside the function since that will get set only once and never updated if the communicator changes.

Removing the reference for those methods for vectors, matrices etc will imply changing the methods of all the other implementations in deal.II for consistency.